File: Library_management_system/Services/library.py
#Library class internally manages a dictionary/Dictionary of books with add_book(), borrow_book(), return_book(), search_by_title()
from models.book import Book

class Library:
    def __init__(self):
        self.books = {} # Key : isbn, value: Book object
    
    # Add book
    def add_book(self, book):
        if book.isbn in self.books:
            print(f"Book with ISBN {book.isbn} already exists.")
            return False
        self.books[book.isbn] = book
        # print(f"\nBook {book.title} added to the library.")
        return True
    
    # Borrow book
    def borrow_book(self, isbn):
        if isbn not in self.books:
            print(f"No book found with ISBN {isbn}.")
            return False
        book = self.books[isbn]
        if book.is_borrowed:
            print(f"Book '{book.title}' is already borrowed.")
            return False
        book.is_borrowed = True
        book.borrow_count += 1
        print(f"\nBook '{book.title}' is borrowed.")
        return True
    
    # Return book
    def return_book(self, isbn):
        if isbn not in self.books:
            print(f"No book found with ISBN {isbn}.")
            return False
        book = self.books[isbn]
        if not book.is_borrowed:
            print(f"Book '{book.title}' was not borrowed.")
            return False
        book.is_borrowed = False
        print(f"Book '{book.title}' is returned.")
        return True
    
    # search book by title
    def search_by_title(self, title):
        results = []
        for book in self.books.values():
            if title.lower() in book.title.lower():
                results.append(book)
        return results
    
    # Getting all books
    def get_all_books(self):
        return list(self.books.values())
    
    # getting total books
    def get_total_books(self):
        return len(self.books)
    
    # getting currently borrowed count
    def get_borrow_count(self):
        count = 0
        for book in self.books.values():
            if book.is_borrowed:
                count += 1
        return count

File: CSharp/Library_Management_system/Services/Library.cs
using Models;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Security.Policy;
using System.Text;
using System.Threading.Tasks;

namespace Services
{
    public class Library
    {
        private Dictionary<string, Book> books = new Dictionary<string, Book>();

        // Add Book
        public bool AddBook(Book book)
        {
            if (books.ContainsKey(book.Isbn))
            {
                Console.WriteLine($"Book with ISBN {book.Isbn} already exists.");
                return false;
            }
            books[book.Isbn] = book;
            //Console.WriteLine($"\nBook {book.Title} added to the library.");
            return true;
        }

        // Borrow Book
        public bool BorrowBook(string isbn)
        {
            if (!books.ContainsKey(isbn))
            {
                Console.WriteLine($"No book found with ISBN {isbn}.");
                return false;
            }
            Book book = books[isbn];
            if (book.IsBorrowed)
            {
                Console.WriteLine($"Book '{book.Title}' is already borrowed.");
                return false;
            }
            book.IsBorrowed = true;
            book.BorrowCount++;
            Console.WriteLine($"\nBook '{book.Title}' is borrowed.");
            return true;
        }

        // Return Book
        public bool ReturnBook(string isbn)
        {
            if (!books.ContainsKey(isbn))
            {
                Console.WriteLine($"No book found with ISBN {isbn}.");
                return false;
            }
            Book book = books[isbn];
            if (!book.IsBorrowed)
            {
                Console.WriteLine($"Book {book.Title} was not borrowed.");
                return false;
            }
            book.IsBorrowed = false;
            Console.WriteLine($"Book '{book.Title}' is returned.");
            return true;
        }

        // Search by title
        public List<Book> SearchByTitle(string title)
        {
            List<Book> result = new List<Book>();
            foreach(Book book in books.Values)
            {
                if (book.Title.ToLower().Contains(title.ToLower()))
                {
                    result.Add(book);
                }
            }
            return result;
        }

        // Get all books
        public List<Book> GetAllBooks()
        {
            return new List<Book>(books.Values);
        }

        // Get total books
        public int GetTotalBooks()
        {
            return books.Count;
        }

        // Get currently borrowed book
        public int GetCurrentlyBorrowedCount()
        {
            int count = 0;
            foreach(Book book in books.Values)
            {
                if (book.IsBorrowed)
                {
                    count++;
                }
            }
            return count;
        }
    }
}
